datalad/core/dataset/gitrepo.py

```diff
@@ -0,0 +1,531 @@
+                                          expect_stderr=expect_stderr))
+        if len(lines) > 1:
+            raise AssertionError(
+                "Expected {} to return single line, but it returned {}"
+                .format(["git"] + args, lines))
+        return lines[0]
+
+    def call_git_success(self, args, files=None, expect_stderr=False):
+        """Call git and return true if the call exit code of 0.
+
+        Parameters
+        ----------
+        args : list of str
+          Arguments to pass to `git`.
+        files : list of str, optional
+          File arguments to pass to `git`. The advantage of passing these here
+          rather than as part of `args` is that the call will be split into
+          multiple calls to avoid exceeding the maximum command line length.
+        expect_stderr : bool, optional
+          Standard error is expected and should not be elevated above the DEBUG
+          level.
+
+        Returns
+        -------
+        bool
+        """
+        try:
+            self._call_git(
+                args, files, expect_fail=True, expect_stderr=expect_stderr)
+        except CommandError:
+            return False
+        return True
+
+    def add_fake_dates(self, env):
+        """Add fake dates to `env`.
+
+        Parameters
+        ----------
+        env : dict or None
+            Environment variables.
+
+        Returns
+        -------
+        A dict (copied from env), with date-related environment
+        variables for git and git-annex set.
+        """
+        env = (env if env is not None else environ).copy()
+        if 'DATALAD_FAKE_DATE' in env:
+            # there is already a fake date setup
+            return env
+        # prevent infinite recursion, should this function be called again
+        # inside the next call.
+        environ['DATALAD_FAKE_DATE'] = '1'
+        last_date = list(self.for_each_ref_(
+            fields='committerdate:raw',
+            count=1,
+            pattern='refs/heads',
+            sort="-committerdate",
+        ))
+        del environ['DATALAD_FAKE_DATE']
+
+        if last_date:
+            # Drop the "contextual" timezone, leaving the unix timestamp.  We
+            # avoid :unix above because it wasn't introduced until Git v2.9.4.
+            last_date = last_date[0]['committerdate:raw'].split()[0]
+            seconds = int(last_date)
+        else:
+            seconds = self.config.obtain("datalad.fake-dates-start")
+        seconds_new = seconds + 1
+        date = "@{} +0000".format(seconds_new)
+
+        lgr.debug("Setting date to %s",
+                  time.strftime("%a %d %b %Y %H:%M:%S +0000",
+                                time.gmtime(seconds_new)))
+
+        env["GIT_AUTHOR_DATE"] = date
+        env["GIT_COMMITTER_DATE"] = date
+        env["GIT_ANNEX_VECTOR_CLOCK"] = str(seconds_new)
+        # leave a marker to prevent duplicate processing
+        env["DATALAD_FAKE_DATE"] = date
```

Comment:
Hmm, do you know a scenario where this marker is used?  Testing with the patch below, which should fail if this value makes it to the other side, all `datalad.core` and `datalad.support.tests.test_gitrepo` tests passed on my end when running with `export DATALAD_FAKE__DATES=1`.

```diff
diff --git a/datalad/core/dataset/gitrepo.py b/datalad/core/dataset/gitrepo.py
index 8780ab713..c67133793 100644
--- a/datalad/core/dataset/gitrepo.py
+++ b/datalad/core/dataset/gitrepo.py
@@ -486,6 +486,8 @@ def add_fake_dates(self, env):
         """
         env = (env if env is not None else environ).copy()
         if 'DATALAD_FAKE_DATE' in env:
+            if env['DATALAD_FAKE_DATE'] != '1':
+                assert 0
             # there is already a fake date setup
             return env
         # prevent infinite recursion, should this function be called again
```

If the marker does have an effect in some cases, I'm concerned that it could break the intended behavior of ticking the faked date on each commit-producing operation, but perhaps I'm thinking about it incorrectly.

Reply:
It seems that is a left-over from an attempt of mine to fix something. I will look into removing it.

Reply:
So this is needed. Or rather it is the way I chose to prevent infinite recursion (the next few lines internally lead to a `call_git()` call that leads to `add_fake_dates()` being called again.

Reply:
Could you provide me an example of how to trigger it?  As I mentioned, at least with the iteration of the PR I was looking at, I could not, and I'm worried it's an unintended change in behavior.

Reply:
I am sorry. I was looking at the diff in the comment, and not at the line you commented on. That lines is removed, and was not needed. My statement referred to https://github.com/datalad/datalad/pull/4689/files#diff-bd096260933e4f20f182a597bd90fe7dR477 being needed. If you remove it, it will trigger an infinite recursion. But that was not what motivated your original comment. Sorry for the confusion.

Reply:
Ah, okay great.  Yep, that one I understand.  Thanks.